Option for host managed memory

eust-dfinity opened the pull request:
As per discussion on host managed memory on zulip started by @lostman here is a simple implementation.

To make things least invasive I ended up with two traits - one for allocator and one for actual memory. This way LinearMemory is not forced to know anything about the allocator and consequently does not need to inherit restrictions placed on the allocator (e.g. Sync).

Additionally DefaultAllocator is made public, so it's possible to create a wrapper around the existing allocator/memory (insert some hooks) and supply it as a custom one.

There are no specific test cases for custom allocators. Can be added if required.
Implementation itself is tested within existing tests with DefaultAllocator.

@alexcrichton @tschneidereit @peterhuene please have a look

alexcrichton commented on crates/api/src/runtime.rs:
```diff
@@ -309,6 +312,12 @@ impl Config {
     }
+
+    /// Sets a custom memory allocator
+    pub fn with_host_memory(&mut self, allocator: Arc<dyn Allocator>) -> &mut Self {
```
FWIW this is making the `wasmtime_runtime` crate a public dependency of the `wasmtime` crate. We've previously avoided doing this because the `wasmtime_*` family of crate are intended to be far more lower level than the `wasmtime` crate.

That being said we haven't quite had a vector of customization like this before where completely dynamic behavior is needed. I suspect we're going to have these grow over time, but I figured it's worth pointing out that this has a few consequences for the `wasmtime` crate:

* This primarily means that if `wasmtime_runtime` has a major version bump, we have to bump the major version of `wasmtime` as well. It's expected that `wasmtime_runtime` will have *lots* of major version bumps because it's just a bunch of internals. For `wasmtime`, however, I think the idea is that eventually we reach a steady state where we just release new patch/minor versions.

* This is also relatively difficult to use since you'll have to match up, manually, with the version of `wasmtime_runtime` that `wasmtime` is using. 

All that is to say that if we can avoid exposing this degree of a public dependency I think we should. One way, for example, would be to define traits in `wasmtime` as well as in `wasmtime_runtime`. The `wasmtime` crate would then provide its own implementation of the traits in `wasmtime_runtime` for provided types. This is a bit wonky, however, from our perspective and may not be the most trivial to manage.

eust-dfinity replied:
Thank you for the info. I was meaning to ask about it as well.
I'll try to move things around, but first I'd rather come to a consensus on the general picture of how the whole thing should work internally.

There is also a question of MemoryPlan which resides in environ now.

alexcrichton commented on crates/runtime/src/memory.rs:
```diff
@@ -9,9 +9,38 @@ use std::cell::RefCell;
 use wasmtime_environ::{MemoryPlan, MemoryStyle, WASM_MAX_PAGES, WASM_PAGE_SIZE};
 
+/// A memory allocator
+pub trait Allocator: Send + Sync {
```
Could you elaborate a bit on your use case for this? I would sort of expect that to override everything about allocation you'd have imported memories, but it sounds like your use case is also covering situations where the host wants to instantiate modules that define their own memory, but the host still wants to configure how those memories are allocated.

eust-dfinity replied:
Our use case is in fact to control all memories used in modules, not just imported ones. Basically we monitor memory usage and track memory changes. Possible reallocations which may happen in wasmtime right now make it more difficult.

alexcrichton commented on crates/runtime/src/memory.rs:
```diff
@@ -9,9 +9,38 @@ use std::cell::RefCell;
+}
+
+/// A linear memory
+pub trait LinearMemory {
```
In terms of safety *something* here needs to be `unsafe` due to the trickiness to get this all right (and segfaults which happen if there's a bug). I think all traits introduced here (both `Allocator` and `LinearMemory`) will need to be `unsafe trait` declarations since they have a number of safety guarantees to uphold in their implementations.

alexcrichton commented on crates/runtime/src/memory.rs:
```diff
@@ -9,9 +9,38 @@ use std::cell::RefCell;
+/// A default memory allocator used by Wasmtime
+pub struct DefaultAllocator;
+
+impl Allocator for DefaultAllocator {
```
The term "Allocator" feels a bit general here, would it be possible to tweak the naming here to indicate that this isn't like a malloc/free allocator but rather an allocator of linear memories?

eust-dfinity commented:
Thanks for the comments!

As you pointed out Allocator is a too general term. I'll be calling it MemoryCreator here.

As for the safety concerns and possible things that can go wrong.
I distinguish two different entities here. The first one is LinearMemory. This is meant to resemble a simple buffer, not thread safe (but you can hardly expect anything more). In a basic case the buffer can have a bigger preallocated size, and extend it's accessible size on grow(). In general case, it could reallocate itself, but current wasmtime's memories already do that, so it's handled.
The true unsafety part comes from vmmemory call, which basically gives away a pointer, so it's up to wasmtime to give a guarantee, that this pointer will not be used again after grow() is called.
On the host side, it needs to guarantee that it won't do anything with this memory while it's owned by wasmtime, excluding actions inside grow() function. Another thing is alignment and size being a multiple of host page size (but I'm actually not sure if these are hard requirements). I think these are all the restrictions that apply here.

The second entity is the MemoryCreator. This one should be thread safe (Sync). It could just call an mmap, or it could provide the memory from some preallocated pool. The memory creator should conform with MemoryPlan. If it can't it should fail. The biggest problem here is that if MemoryPlan struct changes (gets a new field), the maintainers of a custom MemoryCreator may fail to notice that. 

Currently MemoryPlan is quite a rich struct. Maybe it's a good idea to reorganize it? In particular memory limits and offset_guard_size are the fields which seem like good candidates to be passed to the custom allocator. As for MemoryStyle it seems to me like this is more like wasmtime's internal parameter used by wasmtime's MemoryCreator. The Memory::shared flag worries me most, but it doesn't seem to be used at all... Or we could just pass the required parameters explicitly and not through this struct.


> I think it'd also be good to have something like Memory::new_custom or basically something where you can construct a Memory with a custom allocator and then pass that around to mirror the functionality here as well.

Can you elaborate a bit more on that? I'm not sure I understand. Isn't MemoryCreator::new_memory() basically doing that?

alexcrichton commented:
> I'll be calling it MemoryCreator here.

Sounds reasonable to me!

> As for the safety concerns and possible things that can go wrong.

What you wrote down sounds good to me, thanks for writing that up! You're right yeah that the `vmmemory` method is the one that must be accurate generally, and that's why I'm thinking it'll probably be an `unsafe trait`.

> Maybe it's a good idea to reorganize it? 

This is where I think if we could define an API in the `wasmtime` crate which we then translate to underlying layers it might be a good idea. The MemoryPlan structure is currently pretty raw and it might be best to have a layer on top of that which we specifically design for API consumption rather than JIT internals. (the example here is the `shared` flag which is ignored since it's not implemented in the internals, but we still parse it because "one day we'll take a look at it")

> Can you elaborate a bit more on that? 

A right yeah, I'm thinking of `wasmtime::Memory::new` where you should ideally be able to create a `wasmtime::Memory` with your own custom backing scheme, so you don't have to rely on these custom traits.

(Ideally if we could get away with *only* updating `wasmtime::Memory` that'd be great. Then we wouldn't have to have a global config setting of "how to allocate a wasm-defined memory, you'd only have to concern yourself with creating the right `wasmtime::Memory` structures that conform to your needs)

eust-dfinity commented:
I created appropriate traits in the API.
The mechanism in principle does not change when compared to the previous version.
I tried a few different things here, but this one seems most clean to me.
Modifying only Memory is difficult. Memory and LinearMemory are two different concepts.
As it is now, Memory is more of a struct which represents a (memory) inside a module.
LinearMemory is just a buffer and is way more low-level. I think that mixing them up would just create confusion. 

I ended up not adding Memory::new_custom thing. I don't really have a clear picture in my mind of how this should work. Maybe I'm misunderstanding something after all. The problem here is, that we want to be able to capture *all* memories created by "create_memories" from runtime, and current Memory interface allows to create only one. I guess, if that is what you would like to have there, I can add it like that. Let me know.
For now please take a look and confirm if this satisfies all the crate dependency concerns you had.
Thanks!

alexcrichton commented on crates/api/src/trampoline/memory.rs:
```diff
@@ -31,3 +37,41 @@ pub fn create_handle_with_memory(store: &Store, memory: &MemoryType) -> Result<I
 }
+
+struct LinearMemoryProxy {
+    mem: RefCell<Box<dyn LinearMemory>>,
```
How come this is wrapped in a `RefCell`?

eust-dfinity replied:
This is complementary to:
https://github.com/bytecodealliance/wasmtime/blob/a325b62ade47540b4fe528d0c98e3c7dc1bfc042/crates/runtime/src/memory.rs#L147
Basically it's here because RuntimeLinearMemory::vmmemory() takes &self and not &mut self, but needs to return mut ptr.

alexcrichton commented on crates/api/src/externals.rs:
```diff
@@ -853,3 +853,43 @@ impl Memory {
+    fn as_ptr(&self) -> *const u8;
+
+    /// Return the allocated memory as a mutable pointer to u8.
+    fn as_mut_ptr(&mut self) -> *mut u8;
```
I think it's safe to remove this and have `as_ptr` return `*mut u8` since all linear memories need to support interior mutability anyway.

alexcrichton commented on crates/api/src/externals.rs:
```diff
@@ -853,3 +853,43 @@ impl Memory {
+/// The memory should be page aligned and a multiple of page size.
+/// To prevent possible silent overflows, the memory should be protected by a guard page.
+/// Additionally the safety concerns explained in ['Memory'], for accessing the memory
+/// apply here as well.
```
It might be good to leave a statement here too saying that this is a pretty new and experimental feature and it's recommended to be familiar with the code before using it. I'm not confident personally at least that this is an exhaustive set of what you need to guarantee for usage of this to be safe.

eust-dfinity commented:
I added an example test and updated the rest according to the suggestions.

Btw, maybe it's a good idea to make LinearMemory::grow() mutable? (Same for actual grow in wasmtime_runtime)

alexcrichton commented:
Looking good! I think we'll want to remove `&mut self` entirely from the trait (which would remove the need for the `RefCell` in the translation wrapper. There's so much aliasing going on here I'm not sure we can actually guarantee the `&mut` property needed for growth, so implementations will need to internally hold a `RefCell` or a `Mutex` as appropriate

eust-dfinity commented:
> Looking good! I think we'll want to remove `&mut self` entirely from the trait (which would remove the need for the `RefCell` in the translation wrapper. There's so much aliasing going on here I'm not sure we can actually guarantee the `&mut` property needed for growth, so implementations will need to internally hold a `RefCell` or a `Mutex` as appropriate

Done. Also added a test for grow.

alexcrichton commented:
Thanks!
